### PR TITLE
[SPARK-31577][SQL] Fix case-sensitivity and forward name conflict problems when check name conflicts of CTE relations

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/28318, to make the code more readable, by adding some comments to explain the trick and simplify the code to use a boolean flag instead of 2 string sets.

This PR also fixes various problems:
1. the name check should consider case sensitivity
2. forward name conflicts like `with t as (with t2 as ...), t2 as ...` is not a real conflict and we shouldn't fail.

### Why are the changes needed?

correct the behavior

### Does this PR introduce any user-facing change?

yes, fix the fore-mentioned behaviors.

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/cte-nested.sql.out [hunk 1]
```diff
@@ -64,10 +64,9 @@ WITH
   )
 SELECT * FROM t2
 -- !query schema
-struct<>
+struct<scalarsubquery():int>
 -- !query output
-org.apache.spark.sql.AnalysisException
-Name t is ambiguous in nested CTE. Please set spark.sql.legacy.ctePrecedencePolicy to CORRECTED so that name defined in inner CTE takes precedence. If set it to LEGACY, outer CTE definitions will take precedence. See more details in SPARK-28228.;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/cte-nested.sql.out:70]: I tried this query with Spark 2.4.5 (need to replace `t(c) AS (SELECT 1)` with `t AS (SELECT 1 c)` as Spark 2.4 doesn't support `t(c) syntax`), it fails the parser. So we don't need to fail here.

**Comment by cloud-fan** [comment]
cc @peter-toth @xuanyuanking @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 2]
```diff
@@ -41,34 +43,45 @@ object CTESubstitution extends Rule[LogicalPlan] {
...
                 s"Please set ${LEGACY_CTE_PRECEDENCE_POLICY.key} to CORRECTED so that name " +
                 "defined in inner CTE takes precedence. If set it to LEGACY, outer CTE " +
                 "definitions will take precedence. See more details in SPARK-28228.")
-            } else {
-              cteName
             }
-        }.toSet ++ namesInSubqueries
-        assertNoNameConflictsInCTE(child, outerCTERelationNames, newNames)
-        w.innerChildren.foreach(assertNoNameConflictsInCTE(_, newNames, newNames))
+            assertNoNameConflictsInCTE(relation, newNames)
```
Comment (peter-toth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:73]: Do you think we could use `relation.child` here and drop the `SubqueryAlias` case below?

**Comment by peter-toth** [comment]
That flag is a nice trick so LGTM, I just left a minor note.
The other 2 fixes also look good.

File: sql/core/src/test/resources/sql-tests/results/cte-nested.sql.out (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/cte-nested.sql.out:70]: Got it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 3]
```diff
@@ -41,34 +43,45 @@ object CTESubstitution extends Rule[LogicalPlan] {
...
-        val newNames = relations.map {
-          case (cteName, _) =>
-            if (outerCTERelationNames.contains(cteName)) {
-              throw new AnalysisException(s"Name $cteName is ambiguous in nested CTE. " +
+      case With(child, relations) =>
+        val newNames = mutable.ArrayBuffer.empty[String]
+        newNames ++= outerCTERelationNames
+        relations.foreach {
+          case (name, relation) =>
+            if (startOfQuery && outerCTERelationNames.exists(resolver(_, name))) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:67]: My bad. I missed to check case-sensitivity.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala:73]: good idea! updated.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cloud-fan .

**Comment by dongjoon-hyun** [comment]
All Scala/Java/Python test passed, but it's timeouted R testing. 

I ran the R UT manually.
```
══ testthat results  ═══════════════════════════════════════════════════════════
[ OK: 13 | SKIPPED: 0 | WARNINGS: 0 | FAILED: 0 ]
✔ |  OK F W S | Context
✔ |  11       | binary functions [1.5 s]
✔ |   4       | functions on binary files [1.3 s]
✔ |   2       | broadcast variables [0.3 s]
✔ |   5       | functions in client.R
✔ |  46       | test functions in sparkR.R [4.4 s]
✔ |   2       | include R packages [0.2 s]
✔ |   2       | JVM API [0.1 s]
✔ |  75       | MLlib classification algorithms, except for tree-based algorithms [55.2 s]
✔ |  70       | MLlib clustering algorithms [23.8 s]
✔ |   6       | MLlib frequent pattern mining [1.8 s]
✔ |   8       | MLlib recommendation algorithms [4.6 s]
✔ | 136       | MLlib regression algorithms, except for tree-based algorithms [50.6 s]
✔ |   8       | MLlib statistics algorithms [0.4 s]
✔ |  94       | MLlib tree-based algorithms [40.7 s]
✔ |  29       | parallelize() and collect() [0.4 s]
✔ | 428       | basic RDD functions [15.0 s]
✔ |  39       | SerDe functionality [2.2 s]
✔ |  20       | partitionBy, groupByKey, reduceByKey etc. [2.1 s]
✔ |   4       | functions in sparkR.R
✔ |  16       | SparkSQL Arrow optimization [10.5 s]
✔ |   6       | test show SparkDataFrame when eager execution is enabled. [0.7 s]
✔ | 1177       | SparkSQL functions [106.5 s]
✔ |  42       | Structured Streaming [53.8 s]
✔ |  16       | tests RDD function take() [0.5 s]
✔ |  14       | the textFile() function [1.3 s]
✔ |  46       | functions in utils.R [0.3 s]
✔ |   0     1 | Windows-specific tests
────────────────────────────────────────────────────────────────────────────────
test_Windows.R:22: skip: sparkJars tag in SparkContext
Reason: This test is only for Windows, skipped
────────────────────────────────────────────────────────────────────────────────

══ Results ═════════════════════════════════════════════════════════════════════
Duration: 378.6 s

OK:       2306
Failed:   0
Warnings: 0
Skipped:  1
```

Thank you so much for this fix, @cloud-fan and @peter-toth .
Merged to master/3.0.

**Review by xuanyuanking — COMMENTED** [review]
Late LGTM, thanks for the fixing.

**Pull request closed without merging**